### PR TITLE
enhancements for cloud-init support for containers

### PR DESCRIPTION
**Pull request by europaul**
- add support for gzip encoding for write_files
- add better validation for the user-data payload

### CODE REVIEW

File: pkg/pillar/utils/cloudconfig/cloudconfig.go [hunk 1]
```diff
@@ -1,14 +1,17 @@
 package cloudconfig
 
 import (
+	"compress/gzip"
```
Comment (uncleDecart) [inline pkg/pillar/utils/cloudconfig/cloudconfig.go:4]: Why not "archive/zip" so that we don't bring new dependencies?

Comment (europaul) [reply pkg/pillar/utils/cloudconfig/cloudconfig.go:4]: It is a matter of user's expectations: from what I've seen so far cloud-init supports the `gzip`, `base64` and `plain` [encodings](https://cloudinit.readthedocs.io/en/latest/reference/examples.html#writing-out-arbitrary-files). I haven't seen the `zip` encoding used in cloud-init.
We can add additional support for `zip`, but the option to decode `gzip` should still be available.

File: pkg/pillar/utils/cloudconfig/cloudconfig.go [hunk 2]
```diff
@@ -105,3 +130,27 @@ func WriteFile(log *base.LogObject, file WritableFile, rootPath string) error {
...
 	return nil
 }
+
+func decodeGzipContent(content string) ([]byte, error) {
+	gzipReader, err := gzip.NewReader(strings.NewReader(content))
+	if err != nil {
+		return nil, err
+	}
+	defer gzipReader.Close()
+	contentBytes, err := io.ReadAll(io.LimitReader(gzipReader, MaxDecompressedContentSize))
```
Comment (christoph-zededa) [inline pkg/pillar/utils/cloudconfig/cloudconfig.go:140]:
```suggestion
	contentBytes, err := io.ReadAll(io.LimitReader(gzipReader, MaxDecompressedContentSize+1))
```
I understand the comment https://github.com/lf-edge/eve/pull/3553/files#diff-a5680bb7f3f3a3212b4c6f6f870994611ac0c00d74d6c5f7a1037e4cb5a5e426R38 that `MaxDecompressedContentSize` should still be allowed.

Comment (europaul) [reply pkg/pillar/utils/cloudconfig/cloudconfig.go]: Makes sense. Fixed!

File: pkg/pillar/utils/cloudconfig/cloudconfig.go [hunk 3]
```diff
@@ -68,16 +81,28 @@ func WriteFile(log *base.LogObject, file WritableFile, rootPath string) error {
...
+	case "gzip", "gz":
+		contentBytes, err = decodeGzipContent(file.Content)
+		if err != nil {
+			return err
+		}
+
 	default:
-		return errors.New("unsupported encoding type. Only base64 and plain are supported")
+		return errors.New("unsupported encoding type. Only base64, plain and gzip are supported")
 	}
```
Comment (christoph-zededa) [inline pkg/pillar/utils/cloudconfig/cloudconfig.go:107]:
```suggestion
        case "gz+base64", "gzip+base64", "gz+b64", "gzip+b64":
          contentBytes, err = decodeGzipBase64Content(file.Content)

        case "gzip", "gz":
	        contentBytes, err = decodeGzipContent(file.Content)
	default:
                err = errors.New("unsupported encoding type. Only base64, plain and gzip are supported")
    }
	if err != nil {
                return err
       }
```

**Review by eriknordmark — APPROVED** [review]
LGTM

**Review by milan-zededa — APPROVED** [review]

**Review by rouming — APPROVED** [review]
Good. Merging that.

**Pull request merged**